File: ingestion/historical.py
import requests
import pandas as pd
import os
from datetime import datetime

def fetch_historical_data(symbol: str, start_date: str = None) -> pd.DataFrame:
    """
    Fetch historical stock data for a given symbol from the custom NSE API,
    save it as a CSV in the data folder, and return it as a DataFrame.

    Args:
        symbol (str): Ticker symbol (e.g., "INFY", "TCS").
        start_date (str): Optional. Filter data from this date (YYYY-MM-DD).

    Returns:
        pd.DataFrame: Historical stock data.
    """
    api_url = "https://nse-data-api.onrender.com/historical/"
    params = {"symbol": symbol}

    try:
        response = requests.get(api_url, params=params)
        response.raise_for_status()

        data = response.json().get("data", [])
        df = pd.DataFrame(data)

        if not df.empty and 'date' in df.columns:
            df['date'] = pd.to_datetime(df['date'])

            if start_date:
                start = pd.to_datetime(start_date)
                df = df[df['date'] >= start]

            df = df.sort_values("date").reset_index(drop=True)

            # Ensure data directory exists
            os.makedirs("data", exist_ok=True)
            file_path = os.path.join("data", f"{symbol}_historical_data.csv")
            df.to_csv(file_path, index=False)

        return df

    except requests.exceptions.RequestException as e:
        print(f"Error fetching data: {e}")
        return pd.DataFrame()

File: ingestion/historical.js
const readline = require("readline");
const { NseIndia } = require("stock-nse-india");
const { Parser } = require("json2csv");
const fs = require("fs");
const path = require("path");

const nseIndia = new NseIndia();

// Setup readline interface
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

// Ask for user input
rl.question("Enter the NSE stock symbol (e.g., INFY, TCS, IRCTC): ", async (symbol) => {
  const startDate = new Date("2024-01-01");
  const endDate = new Date("2024-05-05");

  console.log(`Fetching data for ${symbol.toUpperCase()}...`);

  try {
    // Fetch data from the NSE India API
    const data = await nseIndia.getEquityHistoricalData(symbol.toUpperCase(), {
      start: startDate,
      end: endDate,
    });

    // Log the raw data to inspect the structure
    console.log("Raw Data:", JSON.stringify(data, null, 2));

    if (!data || data.length === 0) {
      console.log("No data found for the given symbol or date range.");
      rl.close();
      return;
    }

    // Check the structure of the first data entry
    console.log("First Data Entry:", data[0]);

    const parser = new Parser();
    const csv = parser.parse(data);

    const filename = path.join(__dirname, '..', 'data', `${symbol.toUpperCase()}_historical_data.csv`);

    // Save the CSV to the current directory
    fs.writeFileSync(filename, csv);
    console.log(`✅ Historical data saved to ${filename}`);
  } catch (error) {
    console.error("❌ Error fetching or saving data:", error.message);
  }

  rl.close();
});
